controllers/addresspool_controller.go

```diff
@@ -72,23 +81,52 @@ func (r *AddressPoolReconciler) syncMetalLBAddressPool(instance *metallbv1alpha1
 	data.Data["Addresses"] = instance.Spec.Addresses
 	objs, err := render.RenderDir(AddressPoolManifestPath, &data)
 	if err != nil {
-		return errors.Wrapf(err, "Fail to render address-pool manifest")
+		return fmt.Errorf("Fail to render address-pool manifest %v", err)
 	}
 
 	for _, obj := range objs {
-		if err := controllerutil.SetControllerReference(instance, obj, r.Scheme); err != nil {
-			return errors.Wrapf(err, "Failed to set controller reference to %s %s",
-				obj.GetNamespace(), obj.GetName())
-		}
 		if err := apply.ApplyObject(context.Background(), r.Client, obj); err != nil {
-			err = errors.Wrapf(err, "could not apply (%s) %s/%s", obj.GroupVersionKind(),
-				obj.GetNamespace(), obj.GetName())
+			err = fmt.Errorf("could not apply (%s) %s/%s err %v", obj.GroupVersionKind(),
+				obj.GetNamespace(), obj.GetName(), err)
 		}
 	}
 
 	return err
 }
 
+func (r *AddressPoolReconciler) resyncMetallbAddressPool(req ctrl.Request) error {
+	instanceList := &metallbv1alpha1.AddressPoolList{}
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "config",
+			Namespace: req.Namespace,
+		},
+	}
+
+	// Delete the exiting configMap
+	if err := r.Delete(context.Background(), configMap); err != nil {
```

Comment:
the problem k8s doesn't tell which specific object been deleted that why I have to start from scratch plus customers expecting service disrupting during config change.
Also delete the configmap handle the case where the last resource is removed

Reply:
Agree, but if you keep the logic as:

- Always merge all the address pools you have
- update the configmap only once

It will be way more straightforward and less error prone

Reply:
create already work this way I think this might be major rework at this late point for 4.9 plus doing the recreation of addresspool vs configmap from update prespective its the same load plus updating the addresspool and leaving configmap in tact might cause issues for little window till the configuraton is converged unlike with my change

Reply:
I just feel this is not the right way to do:
Probably deleting the configmap is not much of a deal (but it may cause disruptions where a better logic would not).
Deleting a single address pool may cause 100 configmap rewrites while we could have only one.

@russellb @markdgray thoughts on this? I am afraid I not too experienced in how metallb itself behaves.

Reply:
I asked @squeed also to take a look

Reply:
One another important thing a key design good for config CR was to have the addresspool objects separated so we can get events/rbac per object, merging addresspool objects will negate this key requirement

Reply:
That won't change. There will be multiple address pools as they are now, and they will be subject to the same events / rbac as they are now. The difference is how we merge it.
What I am saying here is that instead of merging one at the time, we should merge all of them together in single pass.

Reply:
I guess in the normal metallb case, the config map would be updated on the fly rather than deleted and recreated. It would seem that this would then be a regression from that case.

Reply:
let me clarify 
1 - if you are doing addresspool CR update , configmap will be updated (no delete here).
2 - the current metallb even with multiple addresspools configured it uses one k8s object which is the configmap, while with CR each addresspool is an object by itself and all those objects build a single configmap object, so its not 1:1.
3 - as mentioned earlier k8s one one addresspool object is deleted it doesn't tell what is deleted it says something is deleted so we have to rebuild the config.
4 - its typical during config update specially when deleting objects in k8s, u will need sometime for the state to be updated to the desired state. 

5-  based on fede suggested optimizaton the recreate path is very optimized  to minimize the down time